spec/ic.did

```diff
@@ -44,6 +44,18 @@ service ic : {
   deposit_cycles : (record {canister_id : canister_id}) -> ();
   raw_rand : () -> (blob);
 
+  // Threshold ECDSA signature
+  get_ecdsa_public_key : (record {
+    canister_id : opt canister_id;
+    derivation_path : vec blob;
+    key_id : text;
```

Comment:
Does that make sense to make `key_id` a variant?

Reply:
And maybe rename? I'd expect `key_id` to name a key (which is confusing as the derivation path does that already), not a scheme?

Reply:
Making it a variant means we'll have to modify the spec every time adding a new key_id.

The purpose is to identify which master key to use, as we might want to support more than one, or more than one type. Such details are not decided at the moment, so leaving it vague seems to the best we can do...

Reply:
So it does not identify the scheme? (“The only supported `key_id` value at the moment is `"secp256k1"`.” certainly sounds like it does.”)

If details are not known, then just leave the field out, and add an optional field later when there is clarity what can be specified and how?

Reply:
So it does not identify the scheme? (“The only supported `key_id` value at the moment is `"secp256k1"`.” certainly sounds like it does.”)

If details are not known, then just leave the field out, and add an optional field later when there is clarity what can be specified and how?

Reply:
Initially it is just "seckp256k1". New key_ids of the same scheme could be named as "secp256k1_1“. If using a different scheme, it could be "seckp256r1". Details are not decided yet, but we'd like to avoid having to add new APIs (or change type definitions) to use them.

Likely registry canister will be a place to look up supported key ids.

Reply:
Another way to put is, it is better to treat key_id is a runtime value instead of a statically defined enum.

Reply:
Maybe use a more helpful name initially that does not like it's defining the scheme? Start with `secp256k1_1` for the first existing master key? Or `20220226_secp256k`?

Reply:
I also thought that the key_id might be referring to the scheme because the only supported one is `secp256k1`. Agree that it should have an name which makes it clear that there may be other `secp256k1` keys in the future.

Reply:
It seems to me we are mixing two things here. The first one is the type of cryptographic scheme, which `secp256k1` seems to indicate, and for which I agree with @marcin-dziadus that a variant is a more natural representation. And I do think that changing the spec for additional versions we may be supporting (like P-256, P-384, ...) is perfectly fine, as long as the change is backward compatible.

The second aspect is supporting multiple keys of the same type. Here it's first unclear to me why the support via the `derivation_path` wouldn't be sufficient to solve that as well – is it because we are only supporting non-hardened derivation? And *if* there is a reason to have multiple "named" keys of the same type, then I don't understand why they should be managed globally instead of per canister.

(Finally: I don't think the registry would be the right place to store such data, but that's not really a discussion for this PR anyway.)

Reply:
You can imagine that different subnets hold different secp256k1 master keys right? We then need a way to express that. We could split scheme and "name" into two fields, which together are the id of the key.

Reply:
We leave open the possibility that we might have more than one master key per curve. Maybe there is even more than a constant number of them, in which case an enum would not suffice.  But the feeling was is that the key_id should just be a arbitrary name that gets mapped somehow by the system so that for a given key_id, we know: which subnets actually host that key, what is the curve type and parameters, etc.

Reply:
Arbitrary name is good, but `secp265k` doesn't look arbitrary :-)

Here is an arbitrary name, if someone needs one: `2022a`

Reply:
We could make `key_id` a record that has
- enum curve_id
- text key_name

this shows we can have multiple keys on the same curve, but does not encode some expectations of the curve id into the name. Wdyt?

Reply:
Definitely better! Why would the caller want or need to specify curve_id, though? (I.e. why not ust `key_name`?)

Reply:
I imagined that only key_id is unique, not the key_name, so you could have key_id1 = (secp256k1, "1"), and key_id2 = (Curve25519, "1"), and they are not equal. 

Why include the curve? I guess we need to have the relation somewhere, we could introduce a separate map somewhere, but that seems more complex to me.

Reply:
> You can imagine that different subnets hold different secp256k1 master keys right? We then need a way to express that.

IMO, the right way of solving this would be to make the functions available through a canister (not a virtual one, an actual canister) running on the ECDSA subnet. If there is a second subnet with ECDSA, then it would just run another canister and the addressing of "which key do you want to use" is via which canister you call.

> We could split scheme and "name" into two fields, which together are the id of the key, would you prefer that?

I think that makes more sense, at least it's closer to the intent that you described.

Reply:
I imagined that the key name _is_ unique across the keys of the IC, and not just across the keys of the IC of a particular curve. What's gained but confusion to re-use names?

But all in all it seems easier and cleaner to just omit this field now, and adding an optional field later if and when we have multiple keys; presumably then we have more clarity on how to identify keys.

(BTW, wouldn't the public key be the natural way to identify a key pair? Or an abridged version of that, i.e. a fingerprint? GPG and SSH seem to do that?)

Reply:
+1 to what Joachim proposes above: why not keep simpler for now?

Reply:
There's the following reason why we would like to have the key_name field already now:
We think that it might be a good idea to first launch threshold ECDSA to the community only with a test key and let people play around with it. This can, e.g., be used with a first Bitcoin feature release that only connects to the Bitcoin test network. This would help us in running the threshold ECDSA feature in a kind of a "beta" release before we make the release with a production key with which people associate actual monetary value. Having a string-based key id would allow us to name it accordingly, like "temporary_test_key" to reduce the chance as much as possible that people use the test key for storing real value, e.g., Bitcoin.
Also, we think that having defaults is not optimal and creates risk. It's better if the application needs to clearly specify which key it uses. Defaults also become increasingly harder in the longer run.

Reply:
Well, a name `temporary_test_key` wouldn't have caused this discussion; `secp265k` (which doesn't look like a key name to me, let alone a temporary one) has, though :-)

Reply:
> IMO, the right way of solving this would be to make the functions available through a canister (not a virtual one, an actual canister) running on the ECDSA subnet. If there is a second subnet with ECDSA, then it would just run another canister and the addressing of "which key do you want to use" is via which canister you call.

I don't really understand how this would work with sharing keys between subnets. If two subnets both have keys with the same name, how am I supposed to know if it's the same key shared between the subnets, or if they are different?

Reply:
Yes, it does! (And while I understand why a developer might want to choose a specific curve, I wonder why a developer needs to be bothered with choosing among one of many key names … but maybe I’ll shut up and _not_ suggest to make that field optional to make life simpler for the majority of users who just want to sign stuff :))